src/caret_analyze/architecture/architecture_loaded.py

```diff
@@ -1151,6 +1158,39 @@ def _remove_duplicated(callback_groups: Sequence[CallbackGroupValue]) \
                 ids_.append(callback_group)
         return ids_
 
+    @staticmethod
+    def _validate_callback_ids(callback_groups: Sequence[CallbackGroupValue]):
+        # validate callback group ids uniqueness.
+        callback_ids: set[str] = set()
+        duplicated: list[str] = []
+        for callback_group in callback_groups:
+            str_callback_ids = ''.join(callback_group.callback_ids)
+            if len(str_callback_ids) == 0:
+                continue
+            if str_callback_ids not in callback_ids:
+                callback_ids.add(str_callback_ids)
+            else:
+                duplicated.append(str_callback_ids)
+        if len(duplicated) > 0:
+            msg = f'Duplicated callback ids. {duplicated}. Use first callback ids only.'
+            raise InvalidReaderError(msg)
```

Comment:
callback_group.callback_ids is a tuple, so when assigning to callback_ids: set[str], test_mypy.py warned me to make it str, so I connected it with str.join, but if callback_ids is a tuple in the first place Good, so I changed it that way.

Reply:
I think removing callback_group due to duplication of callback_ids is not good idea. Also, this code works only when all element of callback_ids are duplicated. I think this is not a robust workaround.

I am considering about removing duplication in lower layer, such as `DataFrameFormatted._build_timer_callbacks()`. I will post suggested code later.

Reply:
Can you try this modification ?

lttng_info.py
```
        # In the case of runtime recording, there are cases where duplicates are recorded.
        # If duplicates are left, the instance cannot be uniquely identified and a warning will
        # be issued, so they should be deleted.
        timers.drop_duplicate()

+        # workaround: remove duplication of callback_object.
+        #   We added this workaround because the duplication of callback_object happened in
+        #   real application.
+        duplicated_callback_objects = timers.drop_duplicated_callback_object()
+        if len(duplicated_callback_objects) > 0:
+            duplicated_callback_objects_str = [f'0x{c:X}' for c in duplicated_callback_objects]
+            logger.warning(
+                'Different timer callbacks have same callback_object value. '
+                'This can be happen because of memory reusal by operating system. '
+                'Duplicated callbacks were removed except first one as a workaround. '
+                'This issue should be fixed in the future. '
+                'duplicated callback_objects = '
+                f'{duplicated_callback_objects_str}')
+
        return timers
```

trace_point_data.py
```
    def drop_duplicate(self) -> None:
        """Remove duplicated rows."""
        self._df.drop_duplicates(inplace=True)

+    def drop_duplicated_callback_object(self) -> list[int]:
+        duplicated = self._df.duplicated(subset='callback_object', keep='first')
+        duplicated_callback_objects: list[int]
+        if duplicated.size > 0:
+            duplicated_series = self._df.loc[duplicated]['callback_object']
+            duplicated_series = duplicated_series.drop_duplicates(keep='first')
+            duplicated_callback_objects = duplicated_series.to_list()
+            self._df = self._df.loc[~duplicated]
+        else:
+            duplicated_callback_objects = []
+        return duplicated_callback_objects
+
    def drop_row(self, index: Sequence[int]) -> None:
        """
        Drop row.
```

test_lttng_info.py
```
        ).convert_dtypes()
        assert timer.df.equals(expect)

+    def test_build_timer_callbacks_df_duplication_workaround(self, caplog):
+        data = Ros2DataModel()
+
+        node_handle = 1
+        timer_handles = [3, 4, 5, 6, 7, 8]
+        rmw_handle = 9
+        period_ns = 10
+        callback_objects = [0x1000, 0x1001, 0x1002, 0x1001, 0x1002, 0x1001]
+        callback_group_addr = 13
+        symbols = [
+            'symbol1', 'symbol2', 'symbol3', 'symbol4', 'symbol5', 'symbol6'
+        ]
+
+        caplog.set_level(WARNING)
+
+        data.add_node(0, node_handle, 0, rmw_handle, 'node1', '/')
+
+        for i in range(len(timer_handles)):
+            data.add_timer(0, timer_handles[i], 0, period_ns)
+            data.add_callback_object(timer_handles[i], 0, callback_objects[i])
+            data.add_timer_node_link(timer_handles[i], 0, node_handle)
+            data.add_callback_symbol(callback_objects[i], 0, symbols[i])
+            data.callback_group_add_timer(
+                callback_group_addr, 0, timer_handles[i]
+            )
+
+        data.finalize()
+
+        caplog.clear()
+        timer = DataFrameFormatted._build_timer_callbacks(data)
+        # caplog.record_tuples is list of tuple. Each tuple is
+        # (logger, level, message). Following assertion makes sure log message
+        # contains duplicated callback_object.
+        assert len(caplog.record_tuples) == 1 and \
+            f'0x{callback_objects[1]:X}' in caplog.record_tuples[0][2] and \
+            f'0x{callback_objects[2]:X}' in caplog.record_tuples[0][2]
+
+        expect = pd.DataFrame.from_dict(
+            [
+                {
+                    'callback_id': f'timer_callback_{callback_objects[0]}',
+                    'callback_object': callback_objects[0],
+                    'node_handle': node_handle,
+                    'timer_handle': timer_handles[0],
+                    'callback_group_addr': callback_group_addr,
+                    'period_ns': period_ns,
+                    'symbol': symbols[0],
+                    'construction_order': 0,
+                },
+                {
+                    'callback_id': f'timer_callback_{callback_objects[1]}',
+                    'callback_object': callback_objects[1],
+                    'node_handle': node_handle,
+                    'timer_handle': timer_handles[1],
+                    'callback_group_addr': callback_group_addr,
+                    'period_ns': period_ns,
+                    'symbol': symbols[1],
+                    'construction_order': 0,
+                },
+                {
+                    'callback_id': f'timer_callback_{callback_objects[2]}',
+                    'callback_object': callback_objects[2],
+                    'node_handle': node_handle,
+                    'timer_handle': timer_handles[2],
+                    'callback_group_addr': callback_group_addr,
+                    'period_ns': period_ns,
+                    'symbol': symbols[2],
+                    'construction_order': 0,
+                },
+            ]
+        ).convert_dtypes()
+        assert timer.df.reset_index(drop=True).equals(expect)
+
    def test_build_timer_control_df(self):
```

Reply:
I changed it to remove duplicates at the lower layer using the method you suggested, and confirmed that it was removed at the XXXLoaded level